utilities/ovs-controller.8.in
.\" -*- nroff -*-
.de IQ
.  br
.  ns
.  IP "\\$1"
..
.TH ovs\-controller 8 "@VERSION@" "Open vSwitch" "Open vSwitch Manual"
.ds PN ovs\-controller
.
.SH NAME
ovs\-controller \- simple OpenFlow controller reference implementation
.
.SH SYNOPSIS
.B ovs\-controller
[\fIoptions\fR] \fImethod\fR \fB[\fImethod\fR]\&...
.
.SH DESCRIPTION
\fBovs\-controller\fR manages any number of remote switches over OpenFlow
protocol, causing them to function as L2 MAC-learning switches or hub.
.PP
\fBovs\-controller\fR controls one or more OpenFlow switches, specified as
one or more of the following OpenFlow connection methods:
.
.RS
.so lib/vconn-passive.man
.so lib/vconn-active.man
.RE
.
.SH OPTIONS
.IP "\fB\-n\fR"
.IQ "\fB\-\-noflow\fR"
By default, \fBovs\-controller\fR sets up a flow in each OpenFlow switch
whenever it receives a packet whose destination is known due through
MAC learning.  This option disables flow setup, so that every packet
in the network passes through the controller.
.IP
This option is most useful for debugging.  It reduces switching
performance, so it should not be used in production.
.
.TP
\fB\-\-max\-idle=\fIsecs\fR|\fBpermanent\fR
Sets \fIsecs\fR as the number of seconds that a flow set up by the
controller will remain in the switch's flow table without any matching
packets being seen.  If \fBpermanent\fR is specified, which is not
recommended, flows will never expire.  The default is 60 seconds.
.IP
This option has no effect when \fB\-n\fR (or \fB\-\-noflow\fR) is in use
(because the controller does not set up flows in that case).
.
.IP "\fB\-H\fR"
.IQ "\fB\-\-hub\fR"
By default, the controller acts as an L2 MAC-learning switch.  This
option changes its behavior to that of a hub that floods packets on
all but the incoming port.
.IP
If \fB\-H\fR (or \fB\-\-hub\fR) and \fB\-n\fR (or \fB\-\-noflow\fR) are used
together, then the cumulative effect is that every packet passes
through the controller and every packet is flooded.
.IP
This option is most useful for debugging.  It reduces switching
performance, so it should not be used in production.
.
.IP "\fB\-w\fR[\fIwildcard_mask\fR]"
.IQ "\fB\-\-wildcards\fR[\fB=\fIwildcard_mask\fR]\fR"
By default, \fBovs\-controller\fR sets up exact-match flows.  This
option allows it to set up wildcarded flows, which may reduce
flow setup latency by causing less traffic to be sent up to the
controller.
.IP
The optional \fIwildcard_mask\fR is an OpenFlow wildcard bitmask in
hexadecimal that specifies the fields to wildcard.  If no
\fIwildcard_mask\fR is specified, the default value 0x2820F0 is used
which specifies L2-only switching and wildcards L3 and L4 fields.
Another interesting value is 0x2000EC, which specifies L3-only
switching and wildcards L2 and L4 fields.
.IP
This option has no effect when \fB\-n\fR (or \fB\-\-noflow\fR) is in use
(because the controller does not set up flows in that case).
.
.IP "\fB\-N\fR"
.IQ "\fB\-\-normal\fR"
By default, \fBovs\-controller\fR directs packets to a particular port
or floods them.  This option causes it to direct non-flooded packets
to the OpenFlow \fBOFPP_NORMAL\fR port.  This allows the switch itself
to make decisions about packet destinations.  Support for
\fBOFPP_NORMAL\fR is optional in OpenFlow, so this option may not well
with some non-Open vSwitch switches.
.
.IP "\fB\-\-mute\fR"
Prevents ovs\-controller from replying to any OpenFlow messages sent
to it by switches.
.IP
This option is only for debugging the Open vSwitch implementation of
``fail open'' mode.  It must not be used in production.
.
.IP "\fB\-q \fIid\fR"
.IQ "\fB\-\-queue=\fIid\fR"
By default, \fBovs\-controller\fR uses the default OpenFlow queue for
sending packets and setting up flows.  Use one of these options,
supplying \fIid\fR as an OpenFlow queue ID as a decimal number, to
instead use that specific queue.
.IP
This option is incompatible with \fB\-N\fR or \fB\-\-normal\fR and
with \fB\-H\fR or \fB\-\-hub\fR.  If more than one is specified then
this option takes precedence.
.IP
This option may be useful for testing or debugging quality of service
setups.
.
.IP "\fB\-Q \fIport-name\fB:\fIqueue-id\fR"
.IP "\fB\-\-port\-queue \fIport-name\fB:\fIqueue-id\fR"
Configures packets received on the port named \fIport-name\fR
(e.g. \fBeth0\fR) to be output on OpenFlow queue ID \fIqueue-id\fR
(specified as a decimal number).  For the specified port, this option
overrides the default specified on \fB\-q\fR or \fB\-\-queue\fR.
.IP
This option may be specified any number of times with different
\fIport-name\fR arguments.
.IP
This option is incompatible with \fB\-N\fR or \fB\-\-normal\fR and
with \fB\-H\fR or \fB\-\-hub\fR.  If more than one is specified then
this option takes precedence.
.IP
This option may be useful for testing or debugging quality of service
setups.
.
.IP "\fB\-\-with\-flows \fIfile\fR"
When a switch connects, push the flow entries as described in
\fIfile\fR.  Each line in \fIfile\fR is a flow entry in the format
described for the \fBadd\-flows\fR command in the \fBFlow Syntax\fR
section of the \fBovs\-ofctl\fR(8) man page.
.IP
Use this option more than once to add flows from multiple files.
.
.SS "Public Key Infrastructure Options"
.so lib/ssl.man
.so lib/ssl-peer-ca-cert.man
.ds DD
.so lib/daemon.man
.so lib/vlog.man
.so lib/unixctl.man
.so lib/common.man
.so so lib/ofp-version.man
.
.SH EXAMPLES
.PP
To bind locally to port 6633 (the default) and wait for incoming
connections from OpenFlow switches:
.IP
\fB% ovs\-controller ptcp:\fR
.PP
In the future, the default port number will change to 6653, which is the
IANA-defined value.
.SH "BUGS"
.PP
Configuring a Citrix XenServer to connect to a particular controller
only points the remote OVSDB management connection to that controller.
It does not also configure OpenFlow connections, because the manager
is expected to do that over the management protocol.
\fBovs\-controller\fR is not an Open vSwitch manager and does not know
how to do that.
.PP
As a stopgap workaround, \fBovs\-vsctl\fR can wait for an OVSDB
connection and set the controller, e.g.:
.IP
\fB% ovs\-vsctl \-t0 \-\-db=pssl: \-\-certificate=cert.pem
\-\-ca\-cert=none \-\-private\-key=privkey.pem
\-\-peer\-ca\-cert=cacert.pem set\-controller ssl:\fIip\fR
.SH "SEE ALSO"
.
.BR ovs\-appctl (8),
.BR ovs\-ofctl (8),
.BR ovs\-dpctl (8)


NEWS
v2.0.0 - xx xxx xxxx
---------------------
    - OpenFlow:
      * Experimental support for OpenFlow 1.1 (in addition to 1.2 and
        1.3, which had experimental support in 1.10).
      * New support for matching outer source and destination IP address
        of tunneled packets, for tunnel ports configured with the newly
        added "remote_ip=flow" and "local_ip=flow" options.
      * Support for matching on metadata 'pkt_mark' for interacting with
        other system components. On Linux this corresponds to the skb
        mark.
      * Support matching, rewriting SCTP ports
    - The Interface table in the database has a new "ifindex" column to
      report the interface's OS-assigned ifindex.
    - New "check-oftest" Makefile target for running OFTest against Open
      vSwitch.  See README-OFTest for details.
    - The flow eviction threshold has been moved to the Open_vSwitch table.
    - Database names are now mandatory when specifying ovsdb-server options
      through database paths (e.g. Private key option with the database name
      should look like "--private-key=db:Open_vSwitch,SSL,private_key").
    - Added ovs-dev.py, a utility script helpful for Open vSwitch developers.
    - Support for Linux kernels up to 3.10
    - ovs-ofctl:
      * New "ofp-parse" for printing OpenFlow messages read from a file.
    - Added configurable flow caching support to IPFIX exporter.
    - Dropped support for Linux pre-2.6.32.


v1.11.0 - 28 Aug 2013
---------------------
    - Support for megaflows, which allows wildcarding in the kernel (and
      any dpif implementation that supports wildcards).  Depending on
      the flow table and switch configuration, flow set up rates are
      close to the Linux bridge.
    - The "tutorial" directory contains a new tutorial for some advanced
      Open vSwitch features.
    - Stable bond mode has been removed.
    - The autopath action has been removed.
    - New support for the data encapsulation format of the LISP tunnel
      protocol (RFC 6830).  An external control plane or manual flow
      setup is required for EID-to-RLOC mapping.
    - OpenFlow:
      * The "dec_mpls_ttl" and "set_mpls_ttl" actions from OpenFlow
        1.1 and later are now implemented.
      * New "stack" extension for use in actions, to push and pop from
        NXM fields.
      * The "load" and "set_field" actions can now modify the "in_port".  (This
        allows one to enable output to a flow's input port by setting the
        in_port to some unused value, such as OFPP_NONE.)
    - ovs-dpctl:
      * New debugging commands "add-flow", "mod-flow", "del-flow".
      * "dump-flows" now has a -m option to increase output verbosity.
    - In dpif-based bridges, cache action translations, which can improve
      flow set up performance by 80% with a complicated flow table.
    - New syslog format, prefixed with "ovs|", to be easier to filter.
    - RHEL: Removes the default firewall rule that allowed GRE traffic to
      pass through. Any users that relied on this automatic firewall hole
      will have to manually configure it. The ovs-ctl(8) manpage documents
      the "enable-protocol" command that can be used as an alternative.
    - New CFM demand mode which uses data traffic to indicate interface
      liveness.

v1.10.0 - 01 May 2013
---------------------
    - Bridge compatibility support has been removed.  Any uses that
      rely on ovs-brcompatd will have to stick with Open vSwitch 1.9.x
      or adapt to native Open vSwitch support (e.g. use ovs-vsctl instead
      of brctl).
    - The maximum size of the MAC learning table is now configurable.
    - With the Linux datapath, packets for new flows are now queued
      separately on a per-port basis, so it should no longer be
      possible for a large number of new flows arriving on one port to
      prevent new flows from being processed on other ports.
    - ovs-vsctl:
      * Previously ovs-vsctl would retry connecting to the database forever,
        causing it to hang if ovsdb-server was not running.  Now, ovs-vsctl
        only tries once by default (use --retry to try forever).  This change
        means that you may want to remove uses of --timeout to avoid hangs
        in ovs-vsctl calls.
      * Many "ovs-vsctl" database commands now accept an --if-exists option.
        Please refer to the ovs-vsctl manpage for details.
    - OpenFlow:
      - Experimental support for newer versions of OpenFlow.  See
        the "What versions of OpenFlow does Open vSwitch support?"
        question in the FAQ for more details.
      - The OpenFlow "dp_desc" may now be configured by setting the
        value of other-config:dp-desc in the Bridge table.
      - It is possible to request the OpenFlow port number with the
        "ofport_request" column in the Interface table.
      - The NXM flow_removed message now reports the OpenFlow table ID
        from which the flow was removed.
    - Tunneling:
      - New support for the VXLAN tunnel protocol (see the IETF draft here:
        http://tools.ietf.org/html/draft-mahalingam-dutt-dcops-vxlan-03).
      - Tunneling requires the version of the kernel module paired with
        Open vSwitch 1.9.0 or later.
      - Inheritance of the Don't Fragment bit in IP tunnels (df_inherit)
        is no longer supported.
      - Path MTU discovery is no longer supported.
      - CAPWAP tunneling support removed.
      - Tunnels with multicast destination ports are no longer supported.
    - ovs-dpctl:
      - The "dump-flows" and "del-flows" no longer require an argument
        if only one datapath exists.
    - ovs-appctl:
      - New "vlog/disable-rate-limit" and "vlog/enable-rate-limit"
        commands available allow control over logging rate limits.
      - New "dpif/dump-dps", "dpif/show", and "dpif/dump-flows" command
        that mimic the equivalent ovs-dpctl commands.
    - The ofproto library is now responsible for assigning OpenFlow port

utilities/ovs-controller.c
/*
 * Copyright (c) 2008, 2009, 2010, 2011, 2012, 2013 Nicira, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <signal.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "command-line.h"
#include "compiler.h"
#include "daemon.h"
#include "learning-switch.h"
#include "ofp-parse.h"
#include "ofp-version-opt.h"
#include "ofpbuf.h"
#include "openflow/openflow.h"
#include "poll-loop.h"
#include "rconn.h"
#include "simap.h"
#include "stream-ssl.h"
#include "timeval.h"
#include "unixctl.h"
#include "util.h"
#include "vconn.h"
#include "vlog.h"
#include "socket-util.h"
#include "ofp-util.h"

VLOG_DEFINE_THIS_MODULE(controller);

#define MAX_SWITCHES 16
#define MAX_LISTENERS 16

struct switch_ {
    struct lswitch *lswitch;
};

/* -H, --hub: Learn the ports on which MAC addresses appear? */
static bool learn_macs = true;

/* -n, --noflow: Set up flows?  (If not, every packet is processed at the
 * controller.) */
static bool set_up_flows = true;

/* -N, --normal: Use "NORMAL" action instead of explicit port? */
static bool action_normal = false;

/* -w, --wildcard: 0 to disable wildcard flow entries, an OFPFW10_* bitmask to
 * enable specific wildcards, or UINT32_MAX to use the default wildcards. */
static uint32_t wildcards = 0;

/* --max-idle: Maximum idle time, in seconds, before flows expire. */
static int max_idle = 60;

/* --mute: If true, accept connections from switches but do not reply to any
 * of their messages (for debugging fail-open mode). */
static bool mute = false;

/* -q, --queue: default OpenFlow queue, none if UINT32_MAX. */
static uint32_t default_queue = UINT32_MAX;

/* -Q, --port-queue: map from port name to port number. */
static struct simap port_queues = SIMAP_INITIALIZER(&port_queues);

/* --with-flows: Flows to send to switch. */
static struct ofputil_flow_mod *default_flows;
static size_t n_default_flows;
static enum ofputil_protocol usable_protocols;

/* --unixctl: Name of unixctl socket, or null to use the default. */
static char *unixctl_path = NULL;

static void new_switch(struct switch_ *, struct vconn *);
static void parse_options(int argc, char *argv[]);
static void usage(void) NO_RETURN;

int
main(int argc, char *argv[])
{
    struct unixctl_server *unixctl;
    struct switch_ switches[MAX_SWITCHES];
    struct pvconn *listeners[MAX_LISTENERS];
    int n_switches, n_listeners;
    int retval;
    int i;

    proctitle_init(argc, argv);
    set_program_name(argv[0]);
    parse_options(argc, argv);
    signal(SIGPIPE, SIG_IGN);

    if (argc - optind < 1) {
        ovs_fatal(0, "at least one vconn argument required; "
                  "use --help for usage");
    }

    n_switches = n_listeners = 0;
    for (i = optind; i < argc; i++) {
        const char *name = argv[i];
        struct vconn *vconn;

        retval = vconn_open(name, get_allowed_ofp_versions(), DSCP_DEFAULT,
                            &vconn);
        if (!retval) {
            if (n_switches >= MAX_SWITCHES) {
                ovs_fatal(0, "max %d switch connections", n_switches);
            }
            new_switch(&switches[n_switches++], vconn);
            continue;
        } else if (retval == EAFNOSUPPORT) {
            struct pvconn *pvconn;
            retval = pvconn_open(name, get_allowed_ofp_versions(),
                                 DSCP_DEFAULT, &pvconn);
            if (!retval) {
                if (n_listeners >= MAX_LISTENERS) {
                    ovs_fatal(0, "max %d passive connections", n_listeners);
                }
                listeners[n_listeners++] = pvconn;
            }
        }
        if (retval) {
            VLOG_ERR("%s: connect: %s", name, ovs_strerror(retval));
        }
    }
    if (n_switches == 0 && n_listeners == 0) {
        ovs_fatal(0, "no active or passive switch connections");
    }

    daemonize_start();

    retval = unixctl_server_create(unixctl_path, &unixctl);
    if (retval) {
        exit(EXIT_FAILURE);
    }

    daemonize_complete();

    while (n_switches > 0 || n_listeners > 0) {
        /* Accept connections on listening vconns. */
        for (i = 0; i < n_listeners && n_switches < MAX_SWITCHES; ) {
            struct vconn *new_vconn;

            retval = pvconn_accept(listeners[i], &new_vconn);
            if (!retval || retval == EAGAIN) {
                if (!retval) {
                    new_switch(&switches[n_switches++], new_vconn);
                }
                i++;
            } else {
                pvconn_close(listeners[i]);
                listeners[i] = listeners[--n_listeners];
            }
        }

        /* Do some switching work.  . */
        for (i = 0; i < n_switches; ) {
            struct switch_ *this = &switches[i];
            lswitch_run(this->lswitch);
            if (lswitch_is_alive(this->lswitch)) {
                i++;
            } else {
                lswitch_destroy(this->lswitch);
                switches[i] = switches[--n_switches];
            }
        }

        unixctl_server_run(unixctl);

        /* Wait for something to happen. */
        if (n_switches < MAX_SWITCHES) {
            for (i = 0; i < n_listeners; i++) {
                pvconn_wait(listeners[i]);
            }
        }
        for (i = 0; i < n_switches; i++) {
            struct switch_ *sw = &switches[i];
            lswitch_wait(sw->lswitch);
        }
        unixctl_server_wait(unixctl);
        poll_block();
    }

    return 0;
}

static void
new_switch(struct switch_ *sw, struct vconn *vconn)
{
    struct lswitch_config cfg;
    struct rconn *rconn;

    rconn = rconn_create(60, 0, DSCP_DEFAULT, get_allowed_ofp_versions());
    rconn_connect_unreliably(rconn, vconn, NULL);

    cfg.mode = (action_normal ? LSW_NORMAL
                : learn_macs ? LSW_LEARN
                : LSW_FLOOD);
    cfg.wildcards = wildcards;
    cfg.max_idle = set_up_flows ? max_idle : -1;
    cfg.default_flows = default_flows;
    cfg.n_default_flows = n_default_flows;
    cfg.usable_protocols = usable_protocols;
    cfg.default_queue = default_queue;
    cfg.port_queues = &port_queues;
    cfg.mute = mute;
    sw->lswitch = lswitch_create(rconn, &cfg);
}

static void
add_port_queue(char *s)
{
    char *save_ptr = NULL;
    char *port_name;
    char *queue_id;

    port_name = strtok_r(s, ":", &save_ptr);
    queue_id = strtok_r(NULL, "", &save_ptr);
    if (!queue_id) {
        ovs_fatal(0, "argument to -Q or --port-queue should take the form "
                  "\"<port-name>:<queue-id>\"");
    }

    if (!simap_put(&port_queues, port_name, atoi(queue_id))) {
        ovs_fatal(0, "<port-name> arguments for -Q or --port-queue must "
                  "be unique");
    }
}

static void
parse_options(int argc, char *argv[])
{
    enum {
        OPT_MAX_IDLE = UCHAR_MAX + 1,
        OPT_PEER_CA_CERT,
        OPT_MUTE,
        OPT_WITH_FLOWS,
        OPT_UNIXCTL,
        VLOG_OPTION_ENUMS,
        DAEMON_OPTION_ENUMS,
        OFP_VERSION_OPTION_ENUMS
    };
    static const struct option long_options[] = {
        {"hub",         no_argument, NULL, 'H'},
        {"noflow",      no_argument, NULL, 'n'},
        {"normal",      no_argument, NULL, 'N'},
        {"wildcards",   optional_argument, NULL, 'w'},
        {"max-idle",    required_argument, NULL, OPT_MAX_IDLE},
        {"mute",        no_argument, NULL, OPT_MUTE},
        {"queue",       required_argument, NULL, 'q'},
        {"port-queue",  required_argument, NULL, 'Q'},
        {"with-flows",  required_argument, NULL, OPT_WITH_FLOWS},
        {"unixctl",     required_argument, NULL, OPT_UNIXCTL},
        {"help",        no_argument, NULL, 'h'},
        DAEMON_LONG_OPTIONS,
        OFP_VERSION_LONG_OPTIONS,
        VLOG_LONG_OPTIONS,
        STREAM_SSL_LONG_OPTIONS,
        {"peer-ca-cert", required_argument, NULL, OPT_PEER_CA_CERT},
        {NULL, 0, NULL, 0},
    };
    char *short_options = long_options_to_short_options(long_options);

    for (;;) {
        int indexptr;
        char *error;
        int c;

        c = getopt_long(argc, argv, short_options, long_options, &indexptr);
        if (c == -1) {
            break;
        }

        switch (c) {
        case 'H':
            learn_macs = false;
            break;

        case 'n':
            set_up_flows = false;
            break;

        case OPT_MUTE:
            mute = true;
            break;

        case 'N':
            action_normal = true;
            break;

        case 'w':
            wildcards = optarg ? strtol(optarg, NULL, 16) : UINT32_MAX;
            break;

        case OPT_MAX_IDLE:
            if (!strcmp(optarg, "permanent")) {
                max_idle = OFP_FLOW_PERMANENT;
            } else {
                max_idle = atoi(optarg);
                if (max_idle < 1 || max_idle > 65535) {
                    ovs_fatal(0, "--max-idle argument must be between 1 and "
                              "65535 or the word 'permanent'");
                }
            }
            break;

        case 'q':
            default_queue = atoi(optarg);
            break;

        case 'Q':
            add_port_queue(optarg);
            break;

        case OPT_WITH_FLOWS:
            error = parse_ofp_flow_mod_file(optarg, OFPFC_ADD, &default_flows,
                                            &n_default_flows,
                                            &usable_protocols);
            if (error) {
                ovs_fatal(0, "%s", error);
            }
            break;

        case OPT_UNIXCTL:
            unixctl_path = optarg;
            break;

        case 'h':
            usage();

        VLOG_OPTION_HANDLERS
        OFP_VERSION_OPTION_HANDLERS
        DAEMON_OPTION_HANDLERS

        STREAM_SSL_OPTION_HANDLERS

        case OPT_PEER_CA_CERT:
            stream_ssl_set_peer_ca_cert_file(optarg);
            break;

        case '?':
            exit(EXIT_FAILURE);

        default:
            abort();
        }
    }
    free(short_options);

    if (!simap_is_empty(&port_queues) || default_queue != UINT32_MAX) {
        if (action_normal) {
            ovs_error(0, "queue IDs are incompatible with -N or --normal; "
                      "not using OFPP_NORMAL");
            action_normal = false;
        }

        if (!learn_macs) {
            ovs_error(0, "queue IDs are incompatible with -H or --hub; "
                      "not acting as hub");
            learn_macs = true;
        }
    }
}

static void
usage(void)
{
    printf("%s: OpenFlow controller\n"
           "usage: %s [OPTIONS] METHOD\n"
           "where METHOD is any OpenFlow connection method.\n",
           program_name, program_name);
    vconn_usage(true, true, false);
    daemon_usage();
    ofp_version_usage();
    vlog_usage();
    printf("\nOther options:\n"
           "  -H, --hub               act as hub instead of learning switch\n"
           "  -n, --noflow            pass traffic, but don't add flows\n"
           "  --max-idle=SECS         max idle time for new flows\n"
           "  -N, --normal            use OFPP_NORMAL action\n"
           "  -w, --wildcards[=MASK]  wildcard (specified) bits in flows\n"
           "  -q, --queue=QUEUE-ID    OpenFlow queue ID to use for output\n"
           "  -Q PORT-NAME:QUEUE-ID   use QUEUE-ID for frames from PORT-NAME\n"
           "  --with-flows FILE       use the flows from FILE\n"
           "  --unixctl=SOCKET        override default control socket name\n"
           "  -h, --help              display this help message\n"
           "  -V, --version           display version information\n");
    exit(EXIT_SUCCESS);
}
